src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityLargeMiner.java

```diff
@@ -0,0 +1,623 @@
+    private boolean isActive = true;
+    private boolean done = false;
+    private boolean silkTouch = false;
+    private boolean chunkMode = false;
+
+    private final LinkedList<BlockPos> blockPos = new LinkedList<>();
+    private int aRadius;
+    private int pipeY = 0;
+    private boolean invFull = false;
+    private final int tier;
+    private int overclockAmount;
+    private final int chunkRadius;
+    private final int tick;
+    private final int drillingFluidConsumePerTick;
+    private final int fortune;
+    private final String romanNumeralString;
+
+    private static final Cuboid6 PIPE_CUBOID = new Cuboid6(4 / 16.0, 0.0, 4 / 16.0, 12 / 16.0, 1.0, 12 / 16.0);
+
+
+    public MetaTileEntityLargeMiner(ResourceLocation metaTileEntityId, int tier, Material material, int tick, int chunkRadius, int drillingFluidConsumePerTick, int fortune) {
+        super(metaTileEntityId);
+        this.tick = tick;
+        this.material = material;
+        this.tier = tier;
+        this.chunkRadius = chunkRadius;
+        this.aRadius = chunkRadius * 16;
+        this.drillingFluidConsumePerTick = drillingFluidConsumePerTick;
+        this.fortune = fortune;
+        this.romanNumeralString = GTUtility.romanNumeralString(fortune);
+        reinitializeStructurePattern();
+    }
+
+
+    @Override
+    public void invalidateStructure() {
+        super.invalidateStructure();
+        if (isActive)
+            setActive(false);
+    }
+
+    @Override
+    protected void formStructure(PatternMatchContext context) {
+        super.formStructure(context);
+        initializeAbilities();
+    }
+
+    private void initializeAbilities() {
+        this.importFluidHandler = new FluidTankList(true, getAbilities(MultiblockAbility.IMPORT_FLUIDS));
+        this.outputInventory = new ItemHandlerList(getAbilities(MultiblockAbility.EXPORT_ITEMS));
+        this.energyContainer = new EnergyContainerList(getAbilities(MultiblockAbility.INPUT_ENERGY));
+        this.overclockAmount = (int) Math.pow(2, getVoltageTier() - this.tier);
+    }
+
+    public boolean drainEnergy(boolean simulate) {
+        if (energyContainer.getEnergyStored() >= getMaxVoltage() && !invFull && !testForMax()) {
+            if (!simulate)
+                energyContainer.removeEnergy(GTValues.VA[GTUtility.getTierByVoltage(energyContainer.getInputVoltage())]);
+            return true;
+        }
+        return false;
+    }
+
+    public boolean drainFluid(boolean simulate) {
+        FluidStack drillingFluid = DrillingFluid.getFluid(this.drillingFluidConsumePerTick * overclockAmount);
+        FluidStack canDrain = importFluidHandler.drain(drillingFluid, false);
+        if (canDrain != null && canDrain.amount == this.drillingFluidConsumePerTick) {
+            if (!simulate)
+                importFluidHandler.drain(drillingFluid, true);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    protected void updateFormedValid() {
```

Comment:
Doing this logic in updateFormedValid() means that it is only run when the structure is formed, and it is server side. This method is automatically called for multiblocks as well, so it is probably better to use it instead of a more invasive override.